File: app/src/main/java/com/jiahaoliuliu/akami/modelviewpresenter/BaseActivity.java
package com.jiahaoliuliu.akami.modelviewpresenter;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

/**
 * Created by jiahaoliuliu on 7/7/17.
 */

public class BaseActivity extends AppCompatActivity {

    protected BasePresenter mBasePresenter;
    protected Context mContext;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.mContext = this;
    }

    @Override
    protected void onResume() {
        super.onResume();
        mBasePresenter.onViewShown();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mBasePresenter.onViewHidden();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mBasePresenter.onViewDestroyed();
    }

    /**
     * Set the presenter for this activity.
     *
     * @param basePresenter
     */
    public void setPresenter(BasePresenter basePresenter) {
        this.mBasePresenter = basePresenter;
    }
}

File: app/src/main/java/com/jiahaoliuliu/akami/modelviewpresenter/BaseActivity.kt
package com.jiahaoliuliu.akami.modelviewpresenter

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity

open class BaseActivity : AppCompatActivity() {

    var basePresenter: BasePresenter? = null
    lateinit var context: Context

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        context = this
    }

    override fun onResume() {
        super.onResume()
        basePresenter?.onViewShown()
    }

    override fun onPause() {
        basePresenter?.onViewHidden()
        super.onPause()
    }

    override fun onDestroy() {
        basePresenter?.onViewDestroyed()
        super.onDestroy()
    }
}